ocaml/xapi/xapi_xenops.ml

```diff
@@ -695,9 +695,18 @@ module MD = struct
   let vusbs_of_vm ~__context (vmref, vm) =
     vm.API.vM_VUSBs
     |> List.map (fun self -> Db.VUSB.get_record ~__context ~self)
-    (* get attached PUSBs. *)
-    |> List.filter (fun vusb -> vusb.API.vUSB_attached <> Ref.null)
-    |> List.map (fun vusb -> Db.PUSB.get_record ~__context ~self:vusb.API.vUSB_attached)
+    |> List.map (fun vusb -> Db.USB_group.get_record ~__context ~self:vusb.API.vUSB_USB_group)
+    |> List.map
+      (fun usb_group ->
+        let pusb =
+          try
+            List.hd usb_group.API.uSB_group_PUSBs
+          with _ ->
+            raise Api_errors.(Server_error(internal_error,
+              [Printf.sprintf "there is no PUSB associated with the USB_group"]))
```

Comment:
This needs to include the reference of the USB group because otherwise the log is not helpful.

Reply:
There is a pattern here to get the first element in a usb group - this should be factored into its own function such that we have all the error handling only once.

Reply:
good suggestion, but here we need to wirte something to get the reference of USB_group

Reply:
updated.